### PR TITLE
Fix Spiral to output loss similar to Megatron

### PR DESCRIPTION
**Pull request by ttochi**
### Description
In this PR, I made Spiral and Mobius output the same loss as Megatron and fixed a few bugs.
You can then verify that your fixes work correctly by checking that they output the same loss as `JOB_TYPE=interleaving`.
(There may be a slight diff due to the different devices the optimizer works on)

### Commits
- [Removed deepspeed related code that affects correction check](https://github.com/mcrl/Megatron-LM-mcrl/commit/4eb6d17f93c7e10bdc8a28441e1b495e75ba7543)
  - This code block makes the params to be initialized on GPU
  - Megatron initializes params on CPU, then allocate to the GPU
  - Remove this code because it dosen't have a significant impact on `SpiralInitContext`
- [Fix recompute_data_iterator in mobius](https://github.com/mcrl/Megatron-LM-mcrl/commit/b437e039a62b6878d7d50f5fd0c9993f1a5cb7c4)
  - Minor bug fix in mobius
- [Add synchronization to ensure non_blocking offload finished](https://github.com/mcrl/Megatron-LM-mcrl/commit/fca5a9ec28df2484e29ef7986a4b4ae551eaf82b)
  - Without SSO, `optimizer.step` was executed before `offload_grad` completed
  - ~~After `offload_grad(non_blocking=True)`, offload_grad is executed synchronous --> event recorded before tensor copy completed --> `param.grad = None` is executed while grad is copying~~
  - Since `free_grad`, `optimizer.step` are cpu job, these cpu jobs do not wait event
  - Add synchronize() before start optimizer.step
- [Fix to wait p2p_request in compute stream](https://github.com/mcrl/Megatron-LM-mcrl/commit/65610a1c64fdb2463dc6729304b5ddc43fd4e611)
  - Minor bug fix in mobius

### CODE REVIEW

**Comment by gajagajago** [comment]
Megatron inits params of VocabParallelEmbedding, Linear -> on cuda / other -> on cpu

Current init_context.py makes spiral init all params -> on cuda
In spiral, param is freed & empty on init, while spiral_tensor is conditionally filled (can be empty in multi-node case).

I made it this way because I didn't want to spill temporary param on our allocator's shared memory space.
However, I think that doesn't have any negative impact on cpu memory.

Also, for better correctness check, your method that aligns with Megatron method seems better. But few other code lines in init_context.py should be changed. 
e.g., Current _convert_to_spiral_param() assumes all params initialized at GPU
```
        param.spiral_status = SpiralParamStatus.GPU
```

**Comment by gajagajago** [comment]
I thought about removing the free stream, but freeing parameter weights from GPU is a GPU job, so I think we can't.

**Comment by gajagajago** [comment]
I will separate the commits in this PR into smaller ones
